lib/http2/connection.c

```diff
@@ -1144,7 +1148,10 @@ static void on_upgrade_complete(void *_conn, h2o_socket_t *sock, size_t reqsize)
 
     if (conn->_http1_req_input->size > reqsize) {
         size_t remaining_bytes = conn->_http1_req_input->size - reqsize;
-        h2o_buffer_reserve(&sock->input, remaining_bytes);
+        if ((h2o_buffer_try_reserve(&sock->input, remaining_bytes)).base == NULL) {
```

Comment:
I would appreciate it if you could determine if we need to make this change.

Reply:
Gladly. My take is that the upper-bound on H2 upgrade is protected by `limit-request-body`. In fact, we won't even make it to `h2o_http2_handle_upgrade()` if the entity size exceeds the limit, thanks to the entity size validation earlier in the codepath (specifically around entity_readers).

Here's a quick illustration with an H2O proxy that has a small `limit-request-body` setting:

```
> POST / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
> Connection: Upgrade, HTTP2-Settings
> Upgrade: h2c
> HTTP2-Settings: AAMAAABkAARAAAAAAAIAAAAA
> Content-Length: 1175
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
> 
< HTTP/1.1 413 Request Entity Too Large
< Connection: close
< Content-Length: 27
< Server: h2o/2.3.0-DEV@e9799f71
< content-type: text/plain; charset=utf-8
```

Reply:
Done: b275c7af -- I'm now moving on to tracking down unbounded allocation requests throughout the source tree.